app/Services/LdapAd.php

```
@@ -145,9 +145,17 @@ public function ldapLogin(string $username, string $password): User
             throw new Exception('Unable to find user in LDAP directory!');
         }
 
-        return User::where('username', $username)
+        $user = User::where('username', $username)
                 ->whereNull('deleted_at')->where('ldap_import', '=', 1)
                 ->where('activated', '=', '1')->first();
+        /* Above, I could've just done ->firstOrFail() which would've been cleaner, but it would've been miserable to
+           troubleshoot if it ever came up (giving a really generic and untraceable error message)
+        */
+        if (!$user) {
+            throw new Exception("User is either deleted, not activated (can't log in), not from LDAP, or can't be found in database");
```

Are we certain we want to throw an exception here? Additionally, will this be what the user sees when they're trying to login, or...? We don't want that info to be displayed to the logging-in user.

---

The exception is actually okay; if you look above you'll see some other exceptions being thrown for other "can't find the user" types of situations. They all display the same: "Unknown user or bad password".

It turns out that returning `null` was the big problem here as that's a type error (i.e. this function shouldn't be able to return that). When I tried to expand out the types I got into a gigantic yakshave; but with the exception thrown, it worked as expected.